File: estudos/exercicios/python/Question36.py
import threading

n = 3
sem = threading.Semaphore(1)
barrier = threading.Semaphore(0)


count = 0


def ThreadA():
    global count
    print("statement a1")
    sem.acquire()
    count += 1
    print(count)

    if (n == count):
        barrier.release()

    sem.release()
    barrier.acquire()
    barrier.release()


threads = []
for _ in range(10):
    thread = threading.Thread(target=ThreadA)
    threads.append(thread)
    thread.start()

for thread in threads:
    thread.join()

print("count final:", count)

File: estudos/exercicios/java/Question36.java
package estudos.exercicios.java;

import java.util.concurrent.Semaphore;

public class Question36 {
    static int size = 5;
    static Semaphore mutex = new Semaphore(1);
    static Semaphore sem = new Semaphore(0);
    static int cont = 0;

    public static void main(String[] args) {
        Runnable barrier = (() -> {
            try {
                mutex.acquire();
                cont++;
                if (cont == size) {
                    sem.release();
                }
                mutex.release();

                sem.acquire();
                sem.release();
                System.out.println(Thread.currentThread().getName() + " passou pela barreira");

            } catch (Exception e) {
                Thread.currentThread().interrupt();
            }
        });

        int totalThreads = 100;
        for (int i = 0; i < totalThreads; i++) {
            new Thread(barrier, "Thread-" + (i + 1)).start();
        }

    }

}
